packages/open-truss/src/components/OTUqiDataProvider.tsx

```diff
@@ -63,72 +42,22 @@ const UqiDataProvider: BaseOpenTrussComponentV1<z.infer<typeof Props>> = (
 
   if (queryResults === undefined) return <></>
 
-  Object.entries(output).forEach(([k, v]) => {
-    // TODO make sure signals code is putting in signals even if compoennt doesn't declare that as a property
-    const signal = (props as Record<string, any>)[k]
-    if (signal === undefined) {
-      // TODO figure out how we want to handle this
-      console.log('missing signal in data component')
-      return
-    }
-
-    signal as Signal
-
-    const shape = Array.isArray(v) ? v[0] : v
-
-    if (!isShapeValid(shape, queryResults.metadata.columns)) {
-      // TODO handle error better
-      console.log('shape of data from server does not match output of query')
-    }
-
-    const result = Array.isArray(v) ? queryResults.rows : queryResults.rows[0]
-    signal.value = result
-  })
+  for (const signal of output || []) {
+    const shape = getSignalsShapeFromSignal(signal)
+    const parsedResults = parseUqiResult(queryResults)
+    const defaultValue = shape.parse(undefined)
 
-  return children
-}
+    const result = Array.isArray(defaultValue)
+      ? parsedResults
+      : isObject(defaultValue)
+      ? parsedResults[0]
+      : parsedResults[0]?.[signal.yamlName]
```

Comment:
This is clever! I missed this in the last PR, but could you say more about returning single results?

I think I would suggest that we only do array outputs and thus only accept array signals for `output`. It would keep this simpler and more closely aligns this to UQI, which I think always returns multiple results. It also means we don't need the `signal.yamlName` field.

This would also set us up better to in the future let folks do more custom map-reduce functions: the data provider would always pass along an array that can be acted upon.

If it's really helpful to pluck just the first element, I'd suggest we do that by passing a prop to this data provider instead of only through the definition of the signal we're sending the result to.

If we do what I'm saying, your examples could then look like the following, noting that the `resultsObject` is removed as it's equivalent to `resultsArray`:

```yaml
  signals:
    resultsArray: # OTUqiDataProvider returns collection with rows using the shape of the object
      - name: string
        phone: string
        acctbal: number
        comment: string
    # this maps rows to just `[{ name: "row1Name" }, { name: "row2Name" }, ...]`
    name:
      - name: string
```

Reply:
We made this change to support signal shapes that only have one value instead of an array of values. That way you can re-use display components without having to change the component signature or have to introduce another nested frames/frame that returns the first element.

We can change this if there continue to be strong feels ❤️